source/common/router/config_impl.cc

```diff
@@ -1367,18 +1376,23 @@ RouteConstSharedPtr VirtualHostImpl::getRouteFromEntries(const RouteCallback& cb
                                                          const Http::RequestHeaderMap& headers,
                                                          const StreamInfo::StreamInfo& stream_info,
                                                          uint64_t random_value) const {
-  // No x-forwarded-proto header. This normally only happens when ActiveStream::decodeHeaders
-  // bails early (as it rejects a request), so there is no routing is going to happen anyway.
-  const auto* forwarded_proto_header = headers.ForwardedProto();
-  if (forwarded_proto_header == nullptr) {
+  // In the rare case that X-Forwarded-Proto and scheme disagree (say http URL over an HTTPs
+  // connection), force a redirect based on underlying protocol, rather than URL
+  // scheme, so don't force a redirect for a http:// url served over a TLS
+  // connection.
+  // TODO(alyssar, mattklein123) is the goal here to force TLS, or force https:
+  // scheme? Maybe upstreams will get confused if seeing an http url.
```

Comment:
The original intent here (from a very long time ago) was definitely to force TLS, not force the scheme. So I think this is probably correct?